Create dependencies and later delete all applications causes an internal server error

**Steps to reproduce the bug**
- Create applications using `POST /application`
- Create dependencies using `POST /applications-dependency`
- Delete all applications
- Create a new application and try to execute`GET /applications-dependency` you will see an error.

I'm attaching a video to better explain the steps to be done:

https://user-images.githubusercontent.com/2582866/115597332-71749200-a2d9-11eb-8473-e5c6c0688c0c.mp4

This is the log I see in the backend:

```
application-inventory_1     | 2021-04-21 17:38:39,030 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-2) HTTP Request to /application-inventory/applications-dependency?page=0&size=1000 failed, error id: b0bd9977-a7dc-4af9-94e9-96a5996530b1-1: org.jboss.resteasy.spi.UnhandledException: javax.persistence.EntityNotFoundException: Unable to find io.tackle.applicationinventory.entities.Application with id 1
application-inventory_1     |   at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
application-inventory_1     |   at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
application-inventory_1     |   at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:41)
application-inventory_1     |   at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
application-inventory_1     |   at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
application-inventory_1     |   at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
application-inventory_1     |   at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
application-inventory_1     |   at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
application-inventory_1     |   at java.lang.Thread.run(Thread.java:834)
application-inventory_1     |   at org.jboss.threads.JBossThread.run(JBossThread.java:501)
application-inventory_1     |   at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
application-inventory_1     |   at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
application-inventory_1     | Caused by: javax.persistence.EntityNotFoundException: Unable to find io.tackle.applicationinventory.entities.Application with id 1
application-inventory_1     |   at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder$JpaEntityNotFoundDelegate.handleEntityNotFound(FastBootEntityManagerFactoryBuilder.java:193)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:216)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:332)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:108)
application-inventory_1     |   at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:74)
application-inventory_1     |   at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:104)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.fireLoadNoChecks(SessionImpl.java:1186)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1051)
application-inventory_1     |   at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:697)
application-inventory_1     |   at org.hibernate.type.EntityType.resolve(EntityType.java:464)
application-inventory_1     |   at org.hibernate.type.ManyToOneType.resolve(ManyToOneType.java:240)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad$EntityResolver.lambda$static$0(TwoPhaseLoad.java:605)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntityEntryLoadedState(TwoPhaseLoad.java:248)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:182)
application-inventory_1     |   at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntity(TwoPhaseLoad.java:151)
application-inventory_1     |   at org.hibernate.loader.Loader.initializeEntitiesAndCollections(Loader.java:1199)
application-inventory_1     |   at org.hibernate.loader.Loader.processResultSet(Loader.java:1001)
application-inventory_1     |   at org.hibernate.loader.Loader.doQuery(Loader.java:959)
application-inventory_1     |   at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:349)
application-inventory_1     |   at org.hibernate.loader.Loader.doList(Loader.java:2849)
application-inventory_1     |   at org.hibernate.loader.Loader.doList(Loader.java:2831)
application-inventory_1     |   at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2663)
application-inventory_1     |   at org.hibernate.loader.Loader.list(Loader.java:2658)
application-inventory_1     |   at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:506)
application-inventory_1     |   at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:400)
application-inventory_1     |   at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
application-inventory_1     |   at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1414)
application-inventory_1     |   at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1625)
application-inventory_1     |   at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1593)
application-inventory_1     |   at org.hibernate.query.Query.getResultList(Query.java:165)
application-inventory_1     |   at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.list(CommonPanacheQueryImpl.java:239)
application-inventory_1     |   at io.quarkus.hibernate.orm.panache.runtime.PanacheQueryImpl.list(PanacheQueryImpl.java:149)
application-inventory_1     |   at io.tackle.commons.resources.ListFilteredResource.list(ListFilteredResource.java:226)
application-inventory_1     |   at io.tackle.commons.resources.ListFilteredResource.list(ListFilteredResource.java:194)
application-inventory_1     |   at io.tackle.applicationinventory.resources.ApplicationsDependencyListFilteredResource.listHal(ApplicationsDependencyListFilteredResource.java:46)
application-inventory_1     |   at java.lang.reflect.Method.invoke(Method.java:566)
application-inventory_1     |   at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
application-inventory_1     |   at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)
application-inventory_1     |   at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)
application-inventory_1     |   at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
application-inventory_1     |   at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
application-inventory_1     |   ... 17 more
application-inventory_1     | 
```

--- a/src/main/java/io/tackle/applicationinventory/entities/Application.java
+++ b/src/main/java/io/tackle/applicationinventory/entities/Application.java
@@ -1,5 +1,7 @@
 package io.tackle.applicationinventory.entities;
 
+import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+import io.quarkus.panache.common.Parameters;
 import io.tackle.commons.annotations.CheckType;
 import io.tackle.commons.annotations.Filterable;
 import io.tackle.commons.entities.AbstractEntity;
@@ -7,17 +9,17 @@ import org.hibernate.annotations.ResultCheckStyle;
 import org.hibernate.annotations.SQLDelete;
 import org.hibernate.annotations.Where;
 
-import java.util.Set;
-import java.util.HashSet;
-
 import javax.persistence.CollectionTable;
 import javax.persistence.Column;
+import javax.persistence.ElementCollection;
 import javax.persistence.Entity;
 import javax.persistence.JoinColumn;
+import javax.persistence.PreRemove;
 import javax.persistence.Table;
-import javax.persistence.ElementCollection;
 import javax.persistence.UniqueConstraint;
+import java.util.HashSet;
 import java.util.Objects;
+import java.util.Set;
 
 @Entity
 @Table(name = "application")
@@ -42,6 +44,18 @@ public class Application extends AbstractEntity {
     @Column(name = "tag")
     @Filterable(filterName = "tags.tag", check = CheckType.EQUAL)
     public Set<String> tags = new HashSet<>();
+
+    /**
+     * The unidirectional {@link javax.persistence.ManyToOne} associations from {@link ApplicationsDependency}
+     * and the soft delete approach adopted, prevents the "standard" 'on cascade delete' approach
+     * on the FK constraint definition from working.
+     * So the cascade has been implemented here as a pre-remove application task.
+     */
+    @PreRemove
+    public void preRemove() {
+        ApplicationsDependency.list("from_id = :id OR to_id = :id", Parameters.with("id", id))
+                .forEach(PanacheEntityBase::delete);
+    }
 
     /**
      * equals and hashCode methods overridden for being able to use this bean with the {@link org.jgrapht.Graph}

--- a/src/test/java/io/tackle/applicationinventory/resources/ApplicationsDependencyTest.java
+++ b/src/test/java/io/tackle/applicationinventory/resources/ApplicationsDependencyTest.java
@@ -559,5 +559,13 @@ public class ApplicationsDependencyTest extends SecuredResourceTest {
                 .then()
                 .statusCode(200)
                 .body("_embedded.applications-dependency.size()", is(0));
+
+        // delete 3rd application for not altering the other tests
+        given()
+                .pathParam("id", third.id)
+                .when()
+                .delete("/application/{id}")
+                .then()
+                .statusCode(204);
     }
 }

--- a/src/test/java/io/tackle/applicationinventory/resources/ApplicationsDependencyTest.java
+++ b/src/test/java/io/tackle/applicationinventory/resources/ApplicationsDependencyTest.java
@@ -412,4 +412,152 @@ public class ApplicationsDependencyTest extends SecuredResourceTest {
                 .then()
                 .statusCode(204);
     }
+
+    @Test
+    // https://github.com/konveyor/tackle-application-inventory/issues/40
+    public void testDeletingApplicationsCascadeOnDeletingTheirDependencies() {
+        // create 3 applications
+        Application first = new Application();
+        first.name = "first";
+        first.id = Long.valueOf(given()
+                .contentType(ContentType.JSON)
+                .accept(ContentType.JSON)
+                .body(first)
+                .when()
+                .post("/application")
+                .then()
+                .statusCode(201)
+                .extract()
+                .path("id")
+                .toString());
+
+        Application second = new Application();
+        second.name = "second";
+        second.id = Long.valueOf(given()
+                .contentType(ContentType.JSON)
+                .accept(ContentType.JSON)
+                .body(second)
+                .when()
+                .post("/application")
+                .then()
+                .statusCode(201)
+                .extract()
+                .path("id")
+                .toString());
+
+        Application third = new Application();
+        third.name = "third";
+        third.id = Long.valueOf(given()
+                .contentType(ContentType.JSON)
+                .accept(ContentType.JSON)
+                .body(third)
+                .when()
+                .post("/application")
+                .then()
+                .statusCode(201)
+                .extract()
+                .path("id")
+                .toString());
+
+        // add dependencies 1st->2nd, 1st->3rd, 2nd->3rd
+        ApplicationsDependency firstToSecond = new ApplicationsDependency();
+        firstToSecond.from = first;
+        firstToSecond.to = second;
+        firstToSecond.id = Long.valueOf(
+                given()
+                        .contentType(ContentType.JSON)
+                        .accept(ContentType.JSON)
+                        .body(firstToSecond)
+                        .when()
+                        .post(PATH)
+                        .then()
+                        .statusCode(201)
+                        .extract()
+                        .path("id")
+                        .toString());
+
+        ApplicationsDependency firstToThird = new ApplicationsDependency();
+        firstToThird.from = first;
+        firstToThird.to = third;
+        firstToThird.id = Long.valueOf(
+                given()
+                        .contentType(ContentType.JSON)
+                        .accept(ContentType.JSON)
+                        .body(firstToThird)
+                        .when()
+                        .post(PATH)
+                        .then()
+                        .statusCode(201)
+                        .extract()
+                        .path("id")
+                        .toString());
+
+        ApplicationsDependency secondToThird = new ApplicationsDependency();
+        secondToThird.from = second;
+        secondToThird.to = third;
+        secondToThird.id = Long.valueOf(
+                given()
+                        .contentType(ContentType.JSON)
+                        .accept(ContentType.JSON)
+                        .body(secondToThird)
+                        .when()
+                        .post(PATH)
+                        .then()
+                        .statusCode(201)
+                        .extract()
+                        .path("id")
+                        .toString());
+
+        // check there are 3 applications dependencies
+        given()
+                .accept("application/hal+json")
+                .queryParam("from.id", first.id)
+                .queryParam("from.id", second.id)
+                .queryParam("sort", "id")
+                .when()
+                .get(PATH)
+                .then()
+                .statusCode(200)
+                .body("_embedded.applications-dependency.size()", is(3));
+
+        // delete 1st application
+        given()
+                .pathParam("id", first.id)
+                .when()
+                .delete("/application/{id}")
+                .then()
+                .statusCode(204);
+
+        // check there's just one applications dependency
+        given()
+                .accept("application/hal+json")
+                .queryParam("from.id", first.id)
+                .queryParam("from.id", second.id)
+                .queryParam("sort", "id")
+                .when()
+                .get(PATH)
+                .then()
+                .statusCode(200)
+                .body("_embedded.applications-dependency.size()", is(1));
+
+        // delete 2nd application
+        given()
+                .pathParam("id", second.id)
+                .when()
+                .delete("/application/{id}")
+                .then()
+                .statusCode(204);
+
+        // check there are no applications dependencies left
+        given()
+                .accept("application/hal+json")
+                .queryParam("from.id", first.id)
+                .queryParam("from.id", second.id)
+                .queryParam("sort", "id")
+                .when()
+                .get(PATH)
+                .then()
+                .statusCode(200)
+                .body("_embedded.applications-dependency.size()", is(0));
+    }
 }

